akd_core/src/ecvrf/traits.rs

```diff
@@ -76,10 +97,74 @@ pub trait VRFKeyStorage: Clone + Sync + Send {
         version: u64,
     ) -> Result<Proof, VrfError> {
         let key = self.get_vrf_private_key().await?;
+        Self::get_label_proof_with_key(&key, label, stale, version)
+    }
+
+    /// Retrieve the proof for a specific label
+    fn get_label_proof_with_key(
+        key: &VRFPrivateKey,
+        label: &AkdLabel,
+        stale: bool,
+        version: u64,
+    ) -> Result<Proof, VrfError> {
         let hashed_label = crate::utils::get_hash_from_label_input(label, stale, version);
 
         // VRF proof and hash output
         let proof = key.prove(&hashed_label);
         Ok(proof)
     }
+
+    /// Returns the [NodeLabel]s that corresponds to a collection of (label, stale, version) arguments
+    /// with only a single fetch to retrieve the VRF private key from storage.
+    ///
+    /// Note: The stale boolean here is to indicate whether we are getting the [NodeLabel] for a fresh version,
+    /// or a version that we are retiring.
+    async fn get_node_labels(
+        &self,
+        labels: &[(AkdLabel, bool, u64)],
+    ) -> Result<Vec<(AkdLabel, NodeLabel)>, VrfError> {
+        let key = self.get_vrf_private_key().await?;
+
+        #[cfg(feature = "parallel_vrf")]
+        {
+            let mut join_set = tokio::task::JoinSet::new();
+            let labels_vec = labels.to_vec();
+            for (label, stale, version) in labels_vec.into_iter() {
+                let key_ref = key.clone();
+
+                let future = {
+                    async move {
+                        Self::get_node_label_with_key(&key_ref, &label, stale, version)
+                            .map(|ok_result| (label, ok_result))
```

Comment:
So we ignore the errors here but early terminate in the version without parallel VRFs. 

This changes where we actually catch the errors in these different versions and potentially we lose the `username` information (in `directory.rs:149`) if we don't use the parallel version. Can we match the implementations?

One way is this function returns a vector and we can recover the username without the VRF label in the map. In that case, we should also check for the size of the map compared to the input to see whether this is a user-level problem or more general, i.e., we couldn't generate any VRFs for users.

Reply:
Ah so we're not losing the errors, we're mapping a success to add the label, but the error still propogates. The type of an element in the join set is

```rust
impl Future<Output = Result<(AkdLabel, NodeLabel), VrfError>>
```

So once we capture the output of the future with `join_next()`, we then check if the VRF op was successful or not. However looking at it, it's an infallible operation so I'll remove a lot of the `Result<_,_>`'s. What's fallible is the retrieval of the key from storage.

Reply:
Ahh I believe I see your point actually, we're going to bubble up a vrf err but lose which username it occurred on. I think me removing bits that aren't actually fallible will help clean this up, the only VRF error that can occur in computing node labels is the retrieval of the private key, which isn't dependent on username so shouldn't need to be tied to a specific user acct.